### PR TITLE
fix(apigateway): unable to associate RestApi as a route53 target for late bound domains

### LINKED ISSUE #5539: "API does not define a default domain name" for Apigateway alias target
If you manually add a domain name (without passing domainName to the constructor) using `addDomainName` to ApiGateway.RestApi, the readonly "domainName" is `undefined`, and ApiGateway from route 53 target fails with "API does not define a default domain name"

### Reproduction Steps

```ts
  import { ApiGateway } from '@aws-cdk/aws-route53-targets'
  import { RestApi } from '@aws-cdk/aws-apigateway'

  const restApi = new RestApi(stack, 'id', {})
  const certificate = Certificate.fromCertificateArn(
    stack,
    'certificate',
    'arn:aws:acm:xxx'
  )

  restApi.addDomainName('domain-id', {
    domainName: 'domain.name',
    certificate
  })

  new ApiGateway(restApi) // throws "Unhandled rejection Error: API does not define a default domain name"
```

### Error Log

```
Unhandled rejection Error: API does not define a default domain name
    at new ApiGateway 
```

### Environment

  - **CLI Version      :** 1.19.0 (build 5597bbe)
  - **Framework Version:** 1.19
  - **OS               :** Windows 10
  - **Language         :** Typescript

### Other

Easy fix, should move the `this.domainName` assignment outside of the constructor to a private method, and call it either when on "new" or when "addDomainName"

https://github.com/aws/aws-cdk/blob/0f0e2055cab08551bc6e5dfb8a1b6219368263c5/packages/%40aws-cdk/aws-apigateway/lib/restapi.ts#L255-L257

https://github.com/aws/aws-cdk/blob/0f0e2055cab08551bc6e5dfb8a1b6219368263c5/packages/%40aws-cdk/aws-route53-targets/lib/api-gateway-domain-name.ts#L28-L36
--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by netroy**
fixes #5539

----
*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-apigateway/lib/restapi.ts [hunk 1]
```diff
@@ -291,8 +294,8 @@ export class RestApi extends Resource implements IRestApi {
    * @param id The construct id
    * @param options custom domain options
    */
-  public addDomainName(id: string, options: DomainNameOptions): DomainName {
-    return new DomainName(this, id, {
+  public addDomainName(id: string, options: DomainNameOptions): void {
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigateway/lib/restapi.ts]: `addDomainName()` should continue returning the newly created `DomainName`.

File: packages/@aws-cdk/aws-apigateway/lib/restapi.ts [hunk 2]
```diff
@@ -291,8 +294,8 @@ export class RestApi extends Resource implements IRestApi {
...
+    this._domainName = new DomainName(this, id, {
       ...options,
       mapping: this
     });
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigateway/lib/restapi.ts]: This change will break existing customers when used in a specific situation.

Consider this CDK app.

```ts
  import { ApiGateway } from '@aws-cdk/aws-route53-targets';
  import { RestApi } from '@aws-cdk/aws-apigateway';

  const restApi = new RestApi(stack, 'id', {
    domainName: {
      domainName: 'first-domain'
      ...
  });

  restApi.addDomainName('second-domain', {
    domainName: 'second-domain',
    ...
  });

  const domainName = restApi.domainName;
```

Before this change, `domainName` would point to `first-domain`; while, after this change, `domainName` will now point to `second-domain`.

This needs to be fixed.

Further, add another unit test to assert this and update the documentation of `get domainName()` to state this.

Comment (netroy) [reply packages/@aws-cdk/aws-apigateway/lib/restapi.ts]: didn't even think about that. 
Thanks for the review, I'll send an update later.

File: packages/@aws-cdk/aws-apigateway/lib/restapi.ts [hunk 3]
```diff
@@ -212,14 +212,9 @@ export class RestApi extends Resource implements IRestApi {
...
 
-  /**
-   * The domain name mapped to this API, if defined through the `domainName`
-   * configuration prop.
-   */
-  public readonly domainName?: DomainName;
-
   private readonly methods = new Array<Method>();
   private _latestDeployment: Deployment | undefined;
+  private _domainNames: DomainName[] = [];
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigateway/lib/restapi.ts]: Curious to know why you've chosen the approach to store all of `DomainName` even though they're all not needed.

You could instead have stored the first one with a conditional on `addDomainName()`.

Is there some value you see in storing all of them here? I'm tilting towards only storing what's necessary so the amount of state stored is minimal.

Comment (netroy) [reply packages/@aws-cdk/aws-apigateway/lib/restapi.ts]: probably a premature code from my side. but, I assumed that CDK users might want to be able to iterate through all domains associated with an API Gateway, to be able to change things like Security Policy, Endpoint Configuration, & the Certificate associated with the domain. So, this was an attempt to future proof for that possibility.

Comment (nija-at) [reply packages/@aws-cdk/aws-apigateway/lib/restapi.ts]: I'd rather not. Let's make that change when/if it comes. We may end up deciding that there is a better way to do that. 

Let's change this to do only what's necessary here.

**Comment by netroy** [comment]
@nija-at rebased. Can you have a look at it again, please?

**Pull request merged**